doc: update godeltaprof readme.md (#85)
grafana/pyroscope-go: godeltaprof/README.md

# godeltaprof

> godeltaprof is an efficient delta profiler for memory, mutex, block.

Reviewer: Is "block" actually block storage or something else? 
```suggestion
godeltaprof is an efficient delta profiler for memory, mutex, and block.
```

# Why

> In golang, allocation, mutex and block profiles are cumulative - they only grow over time and show allocations happened since the beginning of the running program.

Reviewer: When used in other docs (for example [Golang tutorial](https://www.golangprograms.com/go-language.html)), it looks like Golang is usually capitalized. Is that how you usually use it? 
```suggestion
In Golang, allocation, mutex and block profiles are cumulative. They only grow over time and show allocations that happened since the beginning of the running program.
```

Not only values grow, but the size of the profile itself grows as well. It could grow up to megabytes in size for long-running processes. These megabytes profiles are called huge profiles in this document.

> In many cases it is much more useful to see the difference between two points in time.
> 
> There is a way to do it with the original runtime/pprof package, they call it delta profile, it requires passing seconds argument to pprof endpoint query.

Reviewer: ```suggestion
In many cases, it's more useful to see the differences between two points in time.
You can use the original runtime/pprof package, called a delta profile, to see these differences. 
Using the delta profile requires passing seconds argument to the pprof endpoint query.
```

```
go tool pprof http://localhost:6060/debug/pprof/heap?seconds=30
```

What this does:
1. dump profile p0
2. sleep
3. dump profile p1
4. decompress and parse protobuf p0
5. decompress and parse protobuf p1
6. subtract p0 from p1

> 7. serialize protobuf and compress the result

Reviewer: Adding initial caps for these phrases

> The resulting profile is *usually* much smaller(p0 may be megabytes, while result is usually tens of kilobytes).
> 
> There are number of issues with this approach.
> 1. Heap profile contains both allocation values and inuse values. Inuse values are not cumulative. So inuse values are corrupted by the subtraction.
> Note: it can be fixed if runtime/pprof package would do the following `p0.ScaleN([]float64{-1,-1,0,0})` instead of `p0.Scale(-1)` - that would substract allocation values and zero out inuse values in p0.
> 2. It requires dumping two profiles.
> 3. It produces a lot of allocations putting pressure on GC.

Reviewer: Is "inuse value" normally one word when referring to Golang? In normal grammar, in-use would be hyphenated because it's a compound adjective (in-use (descriptive - adjective) describes values (noun)). 
```suggestion
The resulting profile is *usually* much smaller (`p0` may be megabytes, while result is usually tens of kilobytes).

There are number of issues with this approach:

1. Heap profile contains both allocation values and in-use values. In-use values are not cumulative. In-use values are corrupted by the subtraction.
  **Note:** It can be fixed if runtime/pprof package uses `p0.ScaleN([]float64{-1,-1,0,0})`, instead of `p0.Scale(-1)` - that would subtract allocation values and zero out in-use values in `p0`.
2. It requires dumping two profiles.
3. It produces a lot of allocations putting pressure on GC.
```

Author: @knylander-grafana when I referred to "inuse" I referred to "inuse_space" and "inuse_objects" profling types from Golang pprof 

https://github.com/golang/go/blob/master/src/runtime/pprof/pprof.go#L127

## DataDog's fastdelta

> Another approach found is DataDog's [fastdelta profiler](https://github.com/DataDog/dd-trace-go/blob/30e1406c2cb62af749df03d559853e1d1de0e3bf/profiler/internal/fastdelta/fd.go#L75)
> 
> It improves the runtime/pprof approach by keeping a copy of the previous profile and subtracting the current profile from it.
> It also does it in a more efficient way by using a custom protobuf pprof parser that doesn't allocate as much memory.
> So it is much faster and produces less garbage. Also does not require dumping two profiles.
> However, it still parses huge profiles up to megabytes, just to discard most of it.

Reviewer: ```suggestion
DataDog's [fastdelta profiler](https://github.com/DataDog/dd-trace-go/blob/30e1406c2cb62af749df03d559853e1d1de0e3bf/profiler/internal/fastdelta/fd.go#L75) uses another approach. 

It improves the runtime/pprof approach by keeping a copy of the previous profile and subtracting the current profile from it.
The fastdelta profiler uses a custom protobuf pprof parser that doesn't allocate as much memory.
This approach is more efficient, faster, and produces less garbage. It also doesn't require using two profiles. 
However, the fastdelta profiler still parses huge profiles up to megabytes, just to discard most of it.
```

## godeltaprof

godeltaprof does a similar job but slightly differently.

The main difference is that delta computation happens before serializing any pprof files using `runtime.MemprofileRecord` and `BlockProfileRecord`
This way we don't need to parse huge profiles, we compute the delta on raw records, reject all zeros and serialize and compress the result.
We dont parse huge profiles, we serialize and compress only small delta profiles.

> The source code for godeltaprof is based(forked) on the original [runtime/pprof package](https://github.com/golang/go/tree/master/src/runtime/pprof)  .
> It is modified to include delta computation before serialization and to expose the new endpoints.
> There are other small improvements/benefits:
> - using `github.com/klauspost/compress/gzip` instead of `compress/gzip`
> - optional lazy mappings reading (they don't change over time for most applications)
> - it is a separate package from runtime, so we can update it independently of 

Reviewer: I updated the wording a little. Does the last bullet still make sense to you? 
```suggestion
The source code for godeltaprof is based (forked) on the original [runtime/pprof package](https://github.com/golang/go/tree/master/src/runtime/pprof).
godeltaprof is modified to include delta computation before serialization and to expose the new endpoints.
There are other small improvements and benefits:
- Using `github.com/klauspost/compress/gzip` instead of `compress/gzip`
- Optional lazy mappings reading (they don't change over time for most applications)
- Separate package from runtime, so updated independently 
```

# benchmarks

> I used memory profiles from [pyroscope](https://github.com/grafana/pyroscope) server.

Reviewer: Someone reading wouldn't know who "I" is - what do you think about changing this? 
```suggestion
These benchmarks used memory profiles from the [pyroscope](https://github.com/grafana/pyroscope) server.
```
